enhancements/kubelet/kubelet-node-sizing.md

```diff
@@ -0,0 +1,212 @@
+---
+title: dynamic-node-sizing
+authors:
+  - "@harche"
+reviewers:
+  - "@rphillips"
+approvers:
+  - "@rphillips"
+creation-date: 2021-02-11
+last-updated: 2021-02-11
+status: implementable
+see-also:
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1857446
+replaces:
+superseded-by:
+---
+
+# Kubelet Automatic Node Sizing
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+Kubelet should have an automatic sizing calculation mechanism, which could give kubelet an ability to dynamically set sizing values for memory and cpu system reserved.
+
+Today the sizing values are passed manually to kubelet using `--kube-reserved` and `--system-reserved` flags. Many cloud providers provide reference values for their customers to help them select optimal values based on the node sizes. e.g. [GKE](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#memory_cpu), [AKS](https://docs.microsoft.com/en-us/azure/aks/concepts-clusters-workloads#resource-reservations)
+
+This enhancement proposes a mechanism to automatically determine the optimal sizing values for any node size irrespective of the cloud provider.
+
+## Motivation
+
+Kubelet’s `system reserved` and `kube reserved` play a crucial role in the OOMKilling the resource intensive pods. Without an adequate enough `system reserved` and `kube reserved` we risk freezing the node making it completely unavailable for other pods. 
+
+We have observed that varying the value of `system reserved` and `kube reserved` with respect to the installed capacity of the node helps to deduce optimal values.
+
+Currently, the only way to customize the `system reserved` and `kube reserved` limits is to pre-calculate the values manually prior to Kubelet start.
+
+### Goals
+
+* Enable Kubelet systemd service to determine the value of the `system reserved` automatically during start up. 
+
+### Non-Goals
+
+* For now the systemd service will only be used for calculating the values of `system reserved`. Similar approach can be taken to dynamically fetch the values of other parameters of the kubelet (e.g. `evictionHard`) but they are out of scope of this enhancement. 
+* Strictly from the OpenShift's point of view, we only need to take care of `system reserved`, and not `kube reserve`. Hence this proposal will not deal with generating optimal values for `kube reserve` 
+
+## Proposal
+
+### Installer Changes
+
+If the user wishes to enable `auto node sizing` during the installation of the cluster itself, they will have to indicate that using a field `metadata.autoNodeSizing` in installation configuration. e.g. 
```

Comment:
My first thought is that this should be in the machine pool.

Reply:
We want to have a cluster wide option that's applicable to all nodes. Either all nodes (workers and masters) have this enabled or not. We didn't want user to specify this option for `compute` or `controlPlane` individually.

Reply:
What is the reasoning for that? Is it required that it be applied cluster-wide for some technical reason? Or is it just for user convenience?

Reply:
This enhancement will set the optimal values of the `system-reserved` for the kubelet on the node. Having the optimal value of `system-reserved` makes a huge difference when node is under severe resource crunch, such as pod running with extremely high memory requirements. 

Without the right value of `system-reserved` the risk of node getting into lock up with `NotReady` status increases. You can set the `system-reserved` value only on specific nodes, since it's just a parameter to a kubelet. But this would not be very ideal from the customer support point of view. Customers will see a non-uniform behaviour in performance within their cluster nodes. Nodes with optimal `system-reserve` will perform substantially better than those without optimal values. 

We want to avoid any ambiguity that may arise if customers see their workload run fine on some nodes but not others. This is the reason, we want to make this a cluster-wide option. That way any in future should we receive any customer support ticket related to node lock ups, it's easy for us to understand if the `system-reserve` was set across the cluster properly or not.

Reply:
/cc @rphillips